src/uct/ib/dc/verbs/dc_verbs.c

```diff
@@ -749,6 +774,316 @@ static unsigned uct_dc_verbs_iface_progress(uct_iface_h tl_iface)
+#if IBV_EXP_HW_TM_DC
+    ucs_status_t status;
+
+    if (self->verbs_common.tm.enabled) {
+        status = uct_dc_iface_create_dct(&self->super,
+                                         &self->verbs_common.tm.xrq);
+        if (status != UCS_OK) {
+            return status;
+        }
+        tl_iface->ops.iface_progress = uct_dc_verbs_iface_progress_tm;
+    }
+#endif
+    uct_base_iface_progress_enable(tl_iface,
+                                   UCT_PROGRESS_SEND | UCT_PROGRESS_RECV);
+    return UCS_OK;
+}
+
+static int uct_dc_verbs_iface_is_reachable(const uct_iface_h tl_iface,
+                                           const uct_device_addr_t *dev_addr,
+                                           const uct_iface_addr_t *iface_addr)
+{
+    uct_dc_verbs_iface_t UCS_V_UNUSED *iface =
+                    ucs_derived_of(tl_iface, uct_dc_verbs_iface_t);
+    uct_dc_verbs_iface_addr_t *addr = (uct_dc_verbs_iface_addr_t *)iface_addr;
+    uint8_t my_type = UCT_RC_VERBS_TM_ENABLED(&iface->verbs_common) ?
+                      UCT_RC_VERBS_IFACE_ADDR_TYPE_TM :
+                      UCT_RC_VERBS_IFACE_ADDR_TYPE_BASIC;
+
+    if ((iface_addr != NULL) && (my_type != addr->addr_type)) {
+        return 0;
+    }
+
+    return uct_ib_iface_is_reachable(tl_iface, dev_addr, iface_addr);
+}
+
+static ucs_status_t uct_dc_verbs_iface_get_address(uct_iface_h tl_iface,
+                                                   uct_iface_addr_t *iface_addr)
+{
+    uct_dc_verbs_iface_t UCS_V_UNUSED *iface = ucs_derived_of(tl_iface,
+                                                              uct_dc_verbs_iface_t);
+    uct_dc_verbs_iface_addr_t *addr = (uct_dc_verbs_iface_addr_t *)iface_addr;
+    ucs_status_t status;
+
+    status = uct_dc_iface_get_address(tl_iface, iface_addr);
+    if (status != UCS_OK) {
+        return status;
+    }
+
+    addr->addr_type = UCT_RC_VERBS_TM_ENABLED(&iface->verbs_common) ?
+                      UCT_RC_VERBS_IFACE_ADDR_TYPE_TM :
+                      UCT_RC_VERBS_IFACE_ADDR_TYPE_BASIC;
+    return UCS_OK;
+}
+
+static ucs_status_t
+uct_dc_verbs_iface_event_arm(uct_iface_h tl_iface, unsigned events)
+{
+    uct_dc_verbs_iface_t *iface = ucs_derived_of(tl_iface, uct_dc_verbs_iface_t);
+    uct_rc_iface_t *rc_iface    = &iface->super.super;
+    int arm_rx_solicited, arm_rx_all;
+    ucs_status_t status;
+
+    status = uct_ib_iface_pre_arm(&rc_iface->super);
+    if (status != UCS_OK) {
+        return status;
+    }
+
+    if (events & UCT_EVENT_SEND_COMP) {
+        status = rc_iface->super.ops->arm_tx_cq(&rc_iface->super);
+        if (status != UCS_OK) {
+            return status;
+        }
+    }
+
+    arm_rx_solicited = 0;
+    arm_rx_all       = 0;
+    if (events & UCT_EVENT_RECV_AM) {
+        arm_rx_solicited = 1; /* to wake up on active messages */
+    }
+    if (((events & UCT_EVENT_SEND_COMP) && rc_iface->config.fc_enabled) ||
+         UCT_RC_VERBS_TM_ENABLED(&iface->verbs_common)) {
```

Comment:
need to extract common function e.g uct_rc_iface_common_event_arm(rc_iface, events, force_rx_all)
rc will use force_rx_all=0 and dc will use force_rx_all=1 if TM is enabled